jps-plugin/src/org/intellij/erlang/jps/builder/ErlangModuleBuildOrderBuilder.java

```diff
@@ -0,0 +1,199 @@
+      @Nullable
+      @Override
+      protected String getDirtyElement(@NotNull File file) throws IOException {
+        return isSourceOrHeader(file) ? file.getAbsolutePath() : null;
+      }
+    }.collectDirtyElements(holder);
+
+    List<ErlangTarget> erlangTargets = ContainerUtil.mapNotNull(context.getProjectDescriptor().getBuildTargetIndex().getAllTargets(), new Function<BuildTarget<?>, ErlangTarget>() {
+      @Nullable
+      @Override
+      public ErlangTarget fun(BuildTarget<?> buildTarget) {
+        return buildTarget instanceof ErlangTarget ? (ErlangTarget) buildTarget : null;
+      }
+    });
+    for (ErlangTarget erlangTarget : erlangTargets) {
+      erlangTarget.setBuildOrder(new ErlangModuleBuildOrder());
+    }
+
+
+    if (dirtyErlangFilePaths.isEmpty()) {
+      LOG.debug("There aren't dirty .erl or .hrl files.");
+    }
+    else {
+      LOG.debug("Search dirty modules.");
+      List<String> sortedDirtyModules = getSortedDirtyModules(context, projectBuildOrder, dirtyErlangFilePaths);
+      addFilesToBuiltTarget(context, sortedDirtyModules);
+    }
+
+  }
+
+  @NotNull
+  @Override
+  public String getPresentableName() {
+    return NAME;
+  }
+
+  @Nullable
+  private static ErlangProjectBuildOrder loadProjectBuildOrder(@NotNull CompileContext context) {
+    return readFromXML(context, BUILD_ORDER_FILE_NAME, ErlangProjectBuildOrder.class);
+  }
+
+  private static boolean isSourceOrHeader(@NotNull File file) {
+    return file.getName().endsWith(".erl") || file.getName().endsWith(".hrl");
+  }
+
+  private static List<String> getSortedDirtyModules(@NotNull CompileContext context,
+                                                    @NotNull ErlangProjectBuildOrder projectBuildOrder,
+                                                    @NotNull List<String> dirtyErlangFilePaths) {
+    SortedModuleDependencyGraph graph = new SortedModuleDependencyGraph(projectBuildOrder);
+    Set<String> allDirtyFiles = getAllDirtyFiles(dirtyErlangFilePaths, GraphGenerator.create(graph));
+    return getSortedDirtyModules(projectBuildOrder.myErlangFiles, allDirtyFiles);
+  }
+
+
+  private static void addPrepareDependenciesFailedMessage(@NotNull CompileContext context) {
+    context.processMessage(new CompilerMessage(NAME, BuildMessage.Kind.WARNING, "The project will be fully rebuilt due to errors."));
+  }
+
+  private static Set<String> getAllDirtyFiles(@NotNull List<String> dirtyFiles,
+                                              @NotNull GraphGenerator<String> graph) {
+    Set<String> allDirtyFiles = ContainerUtil.newHashSet();
+    for (String node : dirtyFiles) {
+      if (node != null) {
+        findDirtyFiles(node, graph, allDirtyFiles);
+      }
+    }
+    return allDirtyFiles;
+  }
+
+  private static void findDirtyFiles(@NotNull String filePath,
+                                     @NotNull GraphGenerator<String> graph,
```

Comment:
I will find place of creating 'graph'. Source code of `SortedModuleDependencyGraph`are clearly for understanding. Do you probably want to propose renaming of 'graph' argument?